### PR TITLE
Change mod resolving to use sat4j's partial boolean functionality rather than minisat

### PR DESCRIPTION
**Pull request by AlexIIL**
This then allows us to inspect the links between mods (like dependencies and conflicts) that caused loading to fail.

This depends on some additional functionality in sat4j, PR here: https://github.com/QuiltMC/quilt-loader-sat4j/pull/1

**NOTE: I haven't written any of the logic to turn the resultant list of causes into a readable error yet**, however this should technically still be better than what we had before - even as-is. Personally I'd quite like to remove the simple mode and just have the advanced mode, as then we can bring all of the error message generation together, however that should wait until the advanced mode can generate the same error strings as the simple mode. This might also simplify plugin integration).

NOTE: This is a quilt port of FabricMC/fabric-loader#279

### CODE REVIEW

**Comment by TheGlitch76** [comment]
This PR needs merged before we start working on the QMJ

**Comment by AlexIIL** [comment]
I'm considering splitting off the parts of `ModResolver.findCompatibleSet` out to a separate class (possibly `ModSolver`), since it feels like `ModResolver` is doing two separate tasks ATM (find mods and resolve a set of them). Does that sound like a good idea? (It would have the disadvantage of making it slightly harder to compare the before/after diff though).

**Comment by AlexIIL** [comment]
An explanation of the problem, from the basics. Hopefully this is helpful when reviewing?

# The Task

One of quilt-loader's jobs is to resolve a valid set of mods from all mods on-disk, respecting their requirements and dependencies.

Unlike simpler loading systems (i.e. forge) quilt also allows for loading mods from *inside* other mods. This doesn't sound like it adds much complexity at first, but this has some very important implications:

* If we find a mod, we might not actually load it
* If we find multiple versions of a single modid then that's fine - we'll pick which one to load
(Except for mods at the root of the "mods/" folder).

Both of these mean we have to be a bit more intelligent when deciding which mods to load, since we can't just check for dependencies and duplicates, one at a time, and be done with it.

# How quilt currently solves this

*Currently* we make each mod option a "boolean variable", and then join those variables together with others to form "clauses", according to their dependencies and conflicts. Then, we use sat4j to "solve" this. But wait... what's a "clause"?

# Boolean satisfiability problem (sat)

This is a well-established field, which I only really found out about through fabric-loader. Before I go off and explain how this relates to quilt, note that there's much better explanations (written by people who have tackled this problem for much longer periods of time) available elsewhere on the web, including the [wikipedia page](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem) as a possible starting point, or perhaps a [sat game](http://www.cril.univ-artois.fr/~roussel/satgame/satgame.php?lang=eng).

Instead what follows is a simplified version of the problem, only including the parts directly relevant for quilt-loader.

There are two parts to a sat problem: variables and clauses.

## Variables

A variable is an option which may either be true or false. (Hence the "boolean" in the title :p ).
To make processing these variables possible we represent them in "DIMACS" format: every variable is assigned a positive number, E.G. 4, 6, or 125637.
To represent a "true" value for a variable we make it positive (+4, +6, +125637), and to represent a "false" value we make it negative (-4, -6, etc).

For a more real example we might have this format in a mods folder:

* buildcraft-9.0.1.jar
    * buildcraft-lib-9.0.1.jar
    * buildcraft-robotics-9.0.1.jar
* quilt-std-1.0.jar
    * quilt-std-core-1.0.jar
* modmenu-7.1.6.jar

Then we'd assign each mod id it's own number:

jar file | mod id | number
--|--|--
buildcraft-9.0.1.jar | buildcraft-all | 1
buildcraft-lib-9.0.1.jar | buildcraft-lib | 2
buildcraft-robotics-9.0.1.jar | buildcraft-robotics | 3
quilt-std-1.0.jar | quilt-std | 4
quilt-std-core-1.0.jar | quilt-std-core | 5
modmenu-7.1.6.jar | modmenu | 6

## Clauses

A clause is a list of variables, which relates them together. For example:

+2, -3

That clause has two parts:

* +2: "buildcraft-lib" is loaded.
* -3: "buildcraft-robotics" is *not* loaded.

In order for a clause to be "valid" at least one* of it's conditions must be met, but not zero.
In this case, we'll have validated the clause if:

* buildcraft-lib is loaded, and buildcraft-robotics is loaded. (Then the "+2" would be met, but the "-3" would not)
* buildcraft-lib is loaded, and buildcraft-robotics is *not* loaded. (Then both would be met)
* Neither are loaded. (Then then "-3" would be met).

The only invalid state would be if "buildcraft-robotics" is loaded, but "buildcraft-lib" is *not* loaded.

Hopefully what this clause means is clean: it's a "sat" representation of a "dependency" clause - in this case "buildcraft-robotics" depends on "buildcraft-lib". Since (as far as the dependency is concerned) it doesn't matter if lib is loaded if buildcraft-robotics isn't, but it does matter if lib isn't loaded if robotics is loaded.

## Lots of clauses

A boolean problem is made up of lots of these clauses. A problem can only be satisfied when *all* of the clauses are satisfied.

# Sat4j

Sat4j is a java implementation of a boolean sat "solver": it takes in lots of variables and clauses, and spits out a list of the values which "satisfy"/meet all of those clauses. Sat4j has lots of modules, but quilt only touches 2: core, and partial-boolean.

Sat4j also adds additional functionality for clauses: we can configure how many of the parts of a clause are needed to satisfy it. For example:

* "atLeast" (Does what you'd expect... at least "x" of the elements must match)
* "atMost" (at most the number given, but may be zero. This just negates every element and passes it to "atLeast", which size-"x")
* Plus many, many more.

(Mod loading uses "atMost" when defining mods to ensure only one of each is loaded).

## Core

From what I can tell this is geared towards solving "competative" base sat problems (or... sudoku puzzles?), and as such it has a simple assumption: the list of clauses *has a correct answer*. In other words, no-one will feed it a problem which has no solution. But wait! That's exactly what we have - we have no guarentee that a set of mods in a mods folder is valid. This is basically where the ContradictionException/"Creating Empty clause ?" comes from - sat4j core seems to optimise clauses as they are added, which can result in contradictions. (I'll come back to this).

## Partial-Boolean (pb)

This module does something clever: it uses the "core" module in order to report problematic clauses. From some brief scanning, this might do something similar to adding a unique variable to every clause, which (if set as true) means that particular clause.

Basically, this module reports problematic clauses (or.. dependencies) to us. Which is exactly what we want!

# The current solution

We use core. Not partial-boolean. Which causes the problems outlined above - we don't actually know what clause went wrong. And so:

# My solution.

Use "partial-boolean" rather than "core", in essence.

However I've also added a lot of other stuff:

## Sat4j.pb.DependencyHelper

Quilt currently manages all of the integers itself - using it's own maps. But PB has a new class that helps a lot with this: "DependencyHelper". This uses two classes to represent clauses and variables. I've called variables "LoadOption" (since it could be used for more than just "is this mod loaded" in the future), and "ModLink" for clauses.

**Comment by AlexIIL** [comment]
Ok, this should be up-to-date now, except for the dependency on my own fork of sat4j (which needs the relavent PR to be accepted and built before I can remove most of my changes from build.gradle.

**Comment by TheGlitch76** [comment]
I've released the new version of sat4j, can you update your buildscript?

**Comment by AlexIIL** [comment]
Where? It doesn't look like https://maven.quiltmc.org/#browse/browse:release:org%2Fquiltmc%2Fquilt-loader-sat4j has it - or am I looking in the wrong place?

**Comment by i509VCB** [comment]
That's because the release failed @TheGlitch76: https://github.com/QuiltMC/quilt-loader-sat4j/actions/runs/901190020

**Comment by AlexIIL** [comment]
It builds correctly now, thanks :)

File: src/main/java/org/quiltmc/loader/impl/gui/QuiltStatusTree.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/gui/QuiltStatusTree.java]: Do you mean using "greater than" / "less than" or the parameters?

Comment (TheGlitch76) [reply src/main/java/org/quiltmc/loader/impl/gui/QuiltStatusTree.java]: The class being named `FabricTreeWarningLevel` inside of `QuiltStatusTree`. I'm trying to slowly remove the references to Fabric as we will be diverging more and more as time goes on.

Comment (TheGlitch76) [reply src/main/java/org/quiltmc/loader/impl/gui/QuiltStatusTree.java]: I can actually manage it myself, don't worry about it.

Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/gui/QuiltStatusTree.java]: Oh, ok.

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:857]: My original intention was that we'd use the "fabric" icon to indicate a "fabric.mod.json" based mod, and the quilt icon for "quilt.mod.json" based mods. As QMJ isn't finished quite yet all mods are fabric based, although I do need to move this behaviour to ModLoadOption / ModCandidate rather than hardcode them here.

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:735]: > Is this TODO supposed to be completed.
No, although I'll document this.
> Could returning null cause problems
No - if `describeError` returns null then `fallbackErrorDescription` will be called instead, which produces much worse error messages than I'd like, but is already implemented.

**Comment by AlexIIL** [comment]
Most of the "error handling" needs to be improved in the future, as currently this spits out errors like in [this comment](https://github.com/FabricMC/fabric-loader/pull/279#issuecomment-730758809) (repeated below):

------------
Screenshot from 2020-11-20 01-01-22

Ideally that would generate an error message like the following, however I think it's still usable as-is

    Error involving "LibMultiPart" (lmp) v1.0.0:
        LibMultiPart requires any version of LibBlockAttributes, of which 2 are available:
            v1.2.1, which cannot load as it requires any version of a missing mod "buildcraft".
            v1.0.0, which cannot load as it requires any version of a missing mod "buildcraft".
        You must install any version of "buildcraft".
    (list of mod sources, identical to the screenshot - starting from "Mod 'LibMultiPart' (lmp( v1.0.0 is being loaded from...)

*(end of comment)*

------------

Currently there's a lot of "empty" / unfinished methods (`ModLink.isNode`, `ModLink.getNodesFrom`, `ModLink.getNodesTo`), although it would probably make more sense to remove them until I actually implement proper error message generation.

However all of the implemented methods (except for ModResolver.describeError) are actually finished - `isNode` is technically a property of the class, rather than per-object.

File: src/main/java/org/quiltmc/loader/impl/QuiltLoaderImpl.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/QuiltLoaderImpl.java:154]: I'd rather not require a system property to run tests - since every time someone makes a JUnit test they they would need to add said system property in order to make quilt not crash.

Technically JUnit tests are a bit of an odd case though - since they don't go through the normal launch process (set the game provider, load mods, setup mixins, etc) I'm not sure what a better solution would be?

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:506]: No.

Currently quilt-loader has two error generation paths: the "simple" one (where there are no jij mods, and as such all mods found must be loaded), and the "advanced" one (where we have jij mods and must use sat4j to find a valid mod set). All of those methods are for generating the error messages for the "simple" type. Ideally, I'd like to unify all error message generation, although as I haven't written proper message generation for "advanced" yet (and don't plan to for this PR) I assume this would be left uncompleted.

**Review by TheGlitch76 — APPROVED** [review]
(what I can understand of this code) looks good to me

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (i509VCB) [inline src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:207]: There a reason why this needs to be fully qualified?

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (i509VCB) [inline src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:367]: May be worth wrapping the error at idx = 0 in a mod resolution exception as well so we don't lose the description like below?

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (i509VCB) [inline src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:376]: Hmm what causes this to be thrown? Or does the resolver logic throw it for us?

I don't see any operations specifying how long we have to finish

File: src/main/java/org/quiltmc/loader/impl/util/SystemProperties.java
Comment (i509VCB) [inline src/main/java/org/quiltmc/loader/impl/util/SystemProperties.java:26]: I'd probably rename the field to `DEBUG_MOD_RESOLVING`

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:207]: Sat4j has multiple classes called `SolverFactory`, both with a method called `newLight()`, however only the partial-boolean one works. (So using the full package name avoids accidentality using the wrong one). This is mostly just because I messed up once during development - however I'd be happy to change this to a normal import.

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:367]: Erm, maybe? If there's only a single error then I can't see why we'd need to say explicitly that there's only 1 error. (This is also to ensure it gets displayed in the GUI correctly - otherwise we'd end up with different errors at different indentation levels. However if there's only 1 error then there's only 1 indentation level to worry about - so we might as well *not* include a header error).

File: src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/discovery/ModResolver.java:376]: I do *not* understand sat4j well enough to explain why it can throw `TimeoutException`. From some quick searching however it looks like it can be thrown if we set a timeout (*which we don't ATM*), or if we cancel it manually by calling `ISolver.expireTimeout` (*perhaps we could expose this via a gui button alongside a progress bar in the future?*), *or* if a solver / problem is undefined in some way (sorry, I don't know if this applies to us).

Either way this can be thrown by either `DependencyHelper.hasASolution` and `DependencyHelper.why`.

File: src/main/java/org/quiltmc/loader/impl/util/SystemProperties.java
Comment (AlexIIL) [reply src/main/java/org/quiltmc/loader/impl/util/SystemProperties.java:26]: Sure!

**Comment by AlexIIL** [comment]
@i509VCB, from GreenCord #loader-general:
> Answers seem fine to me

I'll merge this in 14mins (23:20 UTC), unless anyone else has any concerns?

**Comment by i509VCB** [comment]
Alex could you throw a before and after screenshot on here before merge

**Review by i509VCB — APPROVED** [review]
Sure some edges are a bit rough but the result I've seen is a great improvement.

**Comment by AlexIIL** [comment]
This example uses just https://github.com/Ladysnake/Requiem/releases/download/1.7.9/requiem-1.7.9.jar in a mods folder.

Before:
before

After:
after

**Pull request merged**
